R/map.R

```diff
@@ -218,6 +221,27 @@ map_raw <- function(.x, .f, ...) {
   .Call(map_impl, environment(), ".x", ".f", "raw")
 }
 
+
+#' @rdname map
+#' @param .ptype If `NULL`, the default, the output type is the common type
+#'   of the elements of the result. Otherwise, supply a "prototype" giving
+#'   the desired type of output.
+#' @importFrom vctrs vec_c vec_size vec_ptype_common
+#' @export
+map_vec <- function(.x, .f, ..., .ptype = NULL) {
+  out <- map(.x, .f, ...)
+
+  .ptype <- vec_ptype_common(!!!out, .ptype = .ptype)
+  for (i in seq_along(out)) {
+    if (vec_size(out[[i]]) != 1L) {
+      stop_bad_element_vector(out[[i]], i, .ptype, 1L, what = "Result")
+    }
+  }
+
+  vec_c(!!!out, .ptype = .ptype)
+}
```

Comment:
Regarding output names, I think we need to follow the `map_int()` behavior of prioritizing names from the input.

i.e. the invariant is that the output is the same size as the input, with the same names.

``` r
library(purrr)

map_dbl(list(x = 1, y = 2), ~ c(a = .x))
#> x y 
#> 1 2
```

I do this in `slider::slide_vec()` with something like this:

```r
# it assumes the size 1 checks have already occurred
> slider:::vec_simplify
function (x, ptype) 
{
    names <- vec_names(x)
    x <- vec_set_names(x, NULL)
    out <- vec_unchop(x, ptype = ptype)
    vec_set_names(out, names)
}
```

So I don't think we need `name_spec`, we just uphold this predictable invariant

Reply:
Good point.